service/src/main/java/com/codeforcommunity/processor/ReservationProcessorImpl.java

```diff
@@ -0,0 +1,206 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.*;
+
+import com.codeforcommunity.api.IReservationProcessor;
+import com.codeforcommunity.auth.JWTData;
+import com.codeforcommunity.dto.reservation.*;
+import com.codeforcommunity.enums.PrivilegeLevel;
+import com.codeforcommunity.exceptions.*;
+import java.sql.Timestamp;
+import java.util.Optional;
+import org.jooq.DSLContext;
+import org.jooq.generated.tables.records.*;
+
+public class ReservationProcessorImpl implements IReservationProcessor {
+
+  private final DSLContext db;
+
+  public ReservationProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  public Optional<ReservationsRecord> lastAction(int blockId) {
+    return Optional.ofNullable(
+        db.selectFrom(RESERVATIONS)
+            .where(RESERVATIONS.BLOCK_ID.eq(blockId))
+            .orderBy(RESERVATIONS.PERFORMED_AT.desc())
+            .limit(1)
+            .fetchOne());
+  }
+
+  public void basicChecks(int blockId, Integer userId, Integer teamId) {
+    BlocksRecord block = db.selectFrom(BLOCKS).where(BLOCKS.ID.eq(blockId)).fetchOne();
+
+    if (block == null) {
+      throw new BlockDoesNotExistException(blockId);
+    }
+
+    if (userId != null) {
+      UsersRecord user = db.selectFrom(USERS).where(USERS.ID.eq(userId)).fetchOne();
+
+      if (user == null) {
+        throw new UserDoesNotExistException(userId);
+      }
+    }
+
+    if (teamId != null) {
+      TeamsRecord team = db.selectFrom(TEAMS).where(TEAMS.ID.eq(teamId)).fetchOne();
+
+      if (team == null) {
+        throw new TeamDoesNotExistException(teamId);
+      }
+    }
+
+    if (userId != null && teamId != null) {
+      UsersTeamsRecord user_teams =
+          db.selectFrom(USERS_TEAMS)
+              .where(USERS_TEAMS.USER_ID.eq(userId))
+              .and(USERS_TEAMS.TEAM_ID.eq(teamId))
+              .fetchOne();
+
+      if (user_teams == null) {
+        throw new UserNotOnTeamException(userId, teamId);
+      }
+    }
+  }
+
+  @Override
+  public void makeReservation(JWTData userData, MakeReservationRequest makeReservationRequest) {
+    basicChecks(
+        makeReservationRequest.getBlockID(),
+        userData.getUserId(),
+        makeReservationRequest.getTeamID());
+
+    ReservationsRecord reservationsRecord = db.newRecord(RESERVATIONS);
+    reservationsRecord.setBlockId(makeReservationRequest.getBlockID());
+    reservationsRecord.setUserId(userData.getUserId());
+    reservationsRecord.setTeamId(makeReservationRequest.getTeamID());
+    reservationsRecord.setActionType("reserve");
+    reservationsRecord.setPerformedAt(new Timestamp(System.currentTimeMillis()));
+
+    Optional<ReservationsRecord> maybeReservation = lastAction(reservationsRecord.getBlockId());
+
+    // check if block is open (released, uncomplete or no record)
+    if (maybeReservation.isPresent()) {
```

Comment:
I don't love how much logic is stored in these if statements at the end of each of these functions. Can you pull them out into helper functions that have javadoc that explain the conditions for when they throw exceptions? These are the places that are most likely to change if we change any reservation logic so it'll be good to have them already separated and written as clearly as possible